File: 2300. Successful Pairs of Spells and Potions.py
# max( O(m log m), O(n log m) ) time and O(n) space complexity,
# where n = len(spells) and m = len(potions)
class Solution:
    def successfulPairs(self, spells: List[int], potions: List[int], success: int) -> List[int]:
        # O(n) space complexity
        ret = []

        # O(m log m) time complexity
        potions = sorted(potions)
        n = len(potions)
        
        # O(n log m) time complexity
        for spell in spells:

            # normalize success strength with the potion's strength
            limit = math.ceil(success / spell)

            L = bisect.bisect_left(potions, limit)
            ret.append(n - L)

        return ret

File: 2300. Successful Pairs of Spells and Potions.ts
function successfulPairs(spells: number[], potions: number[], success: number): number[] {
    potions.sort((a, b) => a - b);

    return spells.map(spell => {
        let L: number = 0;
        let R: number = potions.length;
        let mid: number = Math.floor((L + R) / 2);

        while (L < R) {
            const strength: number = spell * potions[mid];
            if (strength < success) {
                L = mid + 1;
            } else {
                R = mid;
            };
            mid = Math.floor((L + R) / 2);
        };

        return potions.length - L;
    });
};
